### PR TITLE
feat: Customizeable Semantic Colors

### PR DESCRIPTION
**Pull request by Skarian**
Have really been enjoying using this package in a current project. 

I found it very convenient to be able to select which catppuccin colors were for each semantic category. This will let me serve my client's needs and I imagine such a feature would be very useful for others out there.

Makes some really nice UI.

Have went ahead and implemented it in this PR, let me know what you think of merging.

### CODE REVIEW

**Comment by Liumingxun** [comment]
Thanks for your work! 

In my opinion, certain semantic colors such as `success`, `warning`, or `error` should not be altered by the user; instead, they should follow to the [style guide](https://github.com/catppuccin/catppuccin/blob/main/docs/style-guide.md), or the theme maybe not have the Catppuccin flavor.

And according to the [Design Philosophy](https://github.com/catppuccin/catppuccin?tab=readme-ov-file#-design-philosophy), the colors of `Monochromatic` may not be suitable for use as the `primary` color. I think the `primary` and `accent` semantic color may use `Accent`(rainbow) color only. As for the `secondary` and `neutral` I have no idea, so just remain the same for now.

Summary:
1. `success`, `warning`, `error`, `info` shouldn't be altered.
2. `primary` and `accent` color should only use the colors within the `AccentName`.
3. `secondary` and `neutral` I have no idea yet, might consider using both `AccentName` and `MonochromaticName` colors.

The above is my thought, feel free to discuss it with me! And last, thanks again for your PR, I am glad to improve this repository with you!

**Comment by Skarian** [comment]
Added another commit. Overall, really aligned to the feedback, have made the requisite changes with a few small tweaks.

1. Only `MonochromaticName` values assignable to `neutral`: This seems pretty reasonable, I could not think of a scenario where that should be accent colors
2. I left `info` customizable. The reasoning is that since blue is default assigned, if one wants to use `blue` as the primary color they would not be able to use `info` without there being no distinction. This is my current exact use case for instance.

Thanks for being open to the PR. Let me know what you think.

File: .DS_Store
Comment (42Willow) [inline .DS_Store:1]: This should be removed, and optionally added to `.gitignore`

File: package-lock.json
Comment (42Willow) [inline package-lock.json:1]: This should also be removed, as there is already a pnpm lockfile

File: README.md [hunk 1]
```diff
@@ -54,16 +54,29 @@ module.exports = {
...
       catppuccin("mocha"),
+      // Or you can optionally customize your semantic colors with catppuccin defined color values, details on available values below
+      // Values not explicitly defined will use default values
+      catppuccin("latte", {primary: 'sky', secondary: 'rosewater'}),
     ],
   },
 };
 ```
 
+#### Available Catppuccin Flavors
```
Comment (42Willow) [inline README.md:67]: This should be removed

File: README.md [hunk 2]
```diff
@@ -54,16 +54,29 @@ module.exports = {
...
+- Latte, Frappe, Macchiato, Mocha
+
+#### Customizable Semantic Colors
+- **Optional fields:** primary, secondary, accent, neutral, success, warning, error, info
```
Comment (42Willow) [inline README.md:71]:
```suggestion
- **Optional fields:** primary, secondary, accent, neutral
```

To be honest I'm not sure if secondary or neutral should be able to be customised.

File: src/index.ts [hunk 3]
```diff
@@ -34,28 +35,52 @@ const themeKeys = [
...
-      'neutral': palette.colors.overlay1.hex,
+      'primary': palette.colors[primary].hex,
+      'secondary': palette.colors[secondary].hex,
+      'accent': palette.colors[accent].hex,
+      'neutral': palette.colors[neutral].hex,
       'success': palette.colors.green.hex,
       'warning': palette.colors.yellow.hex,
       'error': palette.colors.red.hex,
-      'info': palette.colors.blue.hex,
+      'info': palette.colors[info].hex,
```
Comment (42Willow) [inline src/index.ts:79]:
```suggestion
      'info': palette.colors.blue.hex,
```

File: src/index.ts [hunk 4]
```diff
@@ -34,28 +35,52 @@ const themeKeys = [
   'warning-content',
 ] as const
 
-function createFlavor(theme: FlavorName, accent?: AccentName): CustomTheme {
+const defaultColorOptions: DefaultColorOptions = {
+  primary: 'lavender',
+  secondary: 'subtext1',
+  accent: 'rosewater',
+  neutral: 'overlay1',
+  info: 'blue',
```
Comment (42Willow) [inline src/index.ts:43]:
```suggestion
```

File: src/index.ts [hunk 5]
```diff
@@ -34,28 +35,52 @@ const themeKeys = [
...
+}
+
+interface DefaultColorOptions {
+  primary: AccentName
+  secondary: AccentName | MonochromaticName
+  accent: AccentName
+  neutral: MonochromaticName
+  info: AccentName
```
Comment (42Willow) [inline src/index.ts:51]:
```suggestion
```

File: src/index.ts [hunk 6]
```diff
@@ -34,28 +35,52 @@ const themeKeys = [
...
+  return { ...defaultColorOptions, ...options }
+}
+
+function createFlavor(theme: FlavorName, customColors?: CustomColorOptions): CustomTheme {
   const palette = flavorEntries.find(([name]) => name === theme)?.[1]
 
   if (!palette)
     throw new Error(`Flavor ${theme} not found!`)
 
+  const { primary, secondary, accent, neutral, info } = getSemanticColors(customColors)
```
Comment (42Willow) [inline src/index.ts:66]:
```suggestion
  const { primary, secondary, accent, neutral } = getSemanticColors(customColors)
```

**Review by 42Willow — CHANGES_REQUESTED** [review]
Note: this is just my opinion

**Comment by sgoudham** [comment]
I personally wouldn't modify `info`, `warning`, or `success` as they should be consistent throughout the organisation. 

Info being `teal` is something that's in the code editors section but mostly for historical reasons as I believe catppuccin/nvim uses that. I much prefer `blue` and would like to keep that here. I'll make sure to update the style guide.

I'm not entirely sure how I feel about the argument of no distinction because in the end, we're just mapping variables to colours. Good and accessible design shouldn't rely solely on colour to differentiate visual elements for the user. 🤔

**Comment by Liumingxun** [comment]
Thanks to all the participants, I made some adjustments and overloaded the API to retain the original usage. This allows users to transition smoothly to the new version. I also kept the custom info color to cover edge cases.

Thanks again everyone.

**Pull request merged**